### DISCUSSION TITLE
Package specific prep/build/... sections

### DISCUSSION THREAD

**Discussion by voxik** [rpm-software-management/rpm#2849] [Ideas]
AFACIT, the sub-package may include `Source` directive. But is there a chance to have also named prep/build/... sections?

BTW wonder what happens with the string appended behind e.g. `prep`. Can I really put there any garbage I like, such as `%prep asl;dkfjsld;fkj`? But I assume this is likely going to change with #2728

**Comment by ffesti** [comment]
Well, sub package definitions are normal preambles that can contain everything the main preamble can. But that doesn't mean those (e.g. `Source:`) directives are somehow attached to that sub package. Global directives are global no matter where they are.

The build scripts also are global. Attaching them to sub packages does not really make sense. While sub packages already exist as an internal structure right after the spec file is parsed they are not looked at until after the build scripts are run.

**Reply by voxik** [reply]
At this stage, I don't really want to constrain myself by `Source` being attached to subpackage or not. But if there is space for [append / prepend scriptlets](https://github.com/rpm-software-management/rpm/pull/2728), then I don't see the package specific scriptlets to being way off.

**Reply by pmatilai** [reply]
Append/prepend is still operating on exactly one script of a type. Adding support for multiple scripts for each step would be a huge amount of work for very little benefit, I don't see that happening. With append/prepend, you can place sub-package specific build scriptlet sections into their own sections though, so you can have a single %if/%endif around it all.

**Reply by voxik** [reply]
> Append/prepend is still operating on exactly one script of a type.

Maybe it could stay like that in practice, if you are talking about the underlying bash script (which to me is just implementation detail I deliberately ignore)

> With append/prepend, you can place sub-package specific build scriptlet sections into their own sections though, so you can have a single %if/%endif around it all.

Possibly, while I can't really imagine it lacking real experience.

**Comment by voxik** [comment]
Just FTR, this is related to the #2847.

What I am possibly about to do is to package Ruby with a few gems into single binary RPM. So my high level idea is to essentially "join" Ruby .spec file with the .spec files of the gems.
